src/components/updatemenus/scrollbox.js

```diff
@@ -0,0 +1,381 @@
+*/
+
+'use strict';
+
+module.exports = ScrollBox;
+
+var d3 = require('d3');
+
+var Color = require('../color');
+var Drawing = require('../drawing');
+
+var Lib = require('../../lib');
+
+/**
+ * Helper class to setup a scroll box
+ *
+ * @class
+ * @param           gd          Plotly's graph div
+ * @param           container   Container to be scroll-boxed (as a D3 selection)
+ * @param {Object}  position
+ * @param {number}  position.l  Left side position (in pixels)
+ * @param {number}  position.t  Top side (in pixels)
+ * @param {number}  position.w  Width (in pixels)
+ * @param {number}  position.h  Height (in pixels)
+ * @param {string}  id          Id for the clip path to implement the scroll box
+ */
+function ScrollBox(gd, container, position, id) {
+    this.gd = gd;
+    this.container = container;
+    this.position = position;
+    this.id = id;
+}
+
+// scroll bar dimensions
+ScrollBox.barWidth = 2;
+ScrollBox.barLength = 20;
+ScrollBox.barRadius = 2;
+ScrollBox.barPad = 1;
+ScrollBox.barColor = '#808BA4';
+
+/**
+ * If needed, setup a clip path and scrollbars
+ *
+ * @method
+ */
+ScrollBox.prototype.enable = function enable() {
+    var fullLayout = this.gd._fullLayout,
+        fullWidth = fullLayout.width,
+        fullHeight = fullLayout.height,
+        graphSize = fullLayout._size;
+
+    // compute position of scroll box
+    var l = this.position.l,
+        w = this.position.w,
+        t = this.position.t,
+        h = this.position.h,
+        boxW = (w <= fullWidth) ? w : fullWidth - graphSize.r - l,
+        boxH = (h <= fullHeight) ? h : fullHeight - graphSize.b - t,
+        boxL, boxR,
+        boxT, boxB;
+
+    var minSize = 2 * (ScrollBox.barLength + 2 * ScrollBox.barPad);
+    if(boxW <= minSize || fullWidth < boxW) boxW = fullWidth / 4;
+    if(boxH <= minSize || fullHeight < boxH) boxH = fullHeight / 4;
+
+    if(0 <= l && l <= fullWidth) {
+        boxL = l;
+        boxR = boxL + boxW;
+    }
+    else {
+        // align left
+        boxL = 0;
+        boxR = boxL + boxW;
+    }
+
+    if(0 <= t && t <= fullHeight) {
+        boxT = t;
+        boxB = boxT + boxH;
+    }
+    else {
+        // align top
+        boxT = 0;
+        boxB = boxT + boxH;
+    }
+
+    if(boxR > fullWidth) {
+        // align right
+        boxR = fullWidth;
+        boxL = boxR - boxW;
+    }
+
+    if(boxB > fullHeight) {
+        // align bottom
+        boxB = fullHeight;
+        boxT = boxB - boxH;
+    }
+
+    this._box = {
+        l: boxL,
+        t: boxT,
+        w: boxW,
+        h: boxH
+    };
+
+    // compute position of horizontal scroll bar
+    var needsHorizontalScrollBar = (w > boxW),
+        hbarW = ScrollBox.barLength + 2 * ScrollBox.barPad,
+        hbarH = ScrollBox.barWidth + 2 * ScrollBox.barPad,
+        hbarL = boxL,
+        hbarT = (boxB + hbarH < fullHeight) ? boxB : fullHeight - hbarH;
+
+    var hbar = this.container.selectAll('rect.scrollbar-horizontal').data(
```

Comment:
If we cut horizontal scrollbars, dropdown menus with `direction: 'left' | 'right'` will be limited to fit within the plot area.

For legends, we could make Scrollbox provide the option to disable the horizontal scrollbar.